scripts/Engineering/gui/engineering_diffraction/tabs/fitting/data_handling/data_model.py

```diff
@@ -141,21 +136,21 @@ def add_log_to_table(self, ws_name, ws, irow):
                run.getProtonCharge(), ws.getTitle()]
         self.write_table_row(ADS.retrieve("run_info"), row, irow)
         # add log data - loop over existing log workspaces not logs in settings as these might have changed
+        runLogs = [l.name for l in run.getLogData()]
         for log in self._log_names:
```

Comment:
If `proton_charge` is not found we can skip this entire loop and jump straight down rather than having a nested if in the loop. Could we use a dict to something like:

```
runLogs = {log.name : full(2, nan) for log in run.getLogData()
# Which gives us a dict like so: {logA: [2, nan], logB: [2, nan], ...etc.}

# Now we can either process the avg or skip the entire block if we don't have a proton charge:
if "proton_charge" in runLogs:
   # arguably this loop could end up in a method like _norm_log_by_proton_charge()
   for log_name in self._log_names: 
       if log in self._log_values[ws_name]:
           k[log_name] = self._log_values[ws_name][log_name]
       elif log_name in runLogs:
           k[log_name] = AverageLogData(ws_name, LogName=log, FixZero=False)
       else:
           logger.warning(f"File {ws.name()} does not contain log {log}")

# Note: outside if block so it always sets something, either our averaged or (2, nan)
for log, avg_and_stderr in runLogs.items():
   self.write_table_row(ADS.retrieve(log), [avg, stdev], irow)
```

Reply:
I take your point, you're correct we do no averaging if proton-charge doesn't exist. The above doesn't quite work as we only want to do this for a handful of logs (in self._log_names) - i.e. we don't need to write NaNs for all the possible log values in `getLogData()`, just the ones that intersect with the logs selected in the settings. I did it this way to avoid doing multiple loops, but of course I have a repeated if statement in there, I have no feeling for what is more efficient (as there are only a few logs I suspect it doesn't matter) - so probably better going for whatever is more readable.

Reply:
Personally it's your call, if you find my version less readable / flexible I'm happy to leave it at is

Reply:
Well I can add in the scenario of proton_charge = 0 that has just been brought to my attention by the instrument scientist on ENGIN-X so I'll make some changes anyway! I will have a go streamlining it.

Reply:
I have added a check that the integrated proton charge is non-zero and taken elements of your code above to make this section more readable (hopefully).